### ISSUE TITLE
taskHistory.bin is constantly increasing its dimensions

### ISSUE THREAD

**Issue by regrog** [gradle/gradle#5178]
### Expected Behavior
`rootProject/.gradle/4.7/taskHistory/taskHistory.bin` should not grow so big: in the order of GB

### Current Behavior
`taskHistory.bin` is slowly but constantly growing up to 2, 5 or 10 GB.
gradle clean doesn't clean this file because it's part of the deamon cache I think.

### Context
I have about 200 java projects with gradle nature, the root project has a gradle wrapper in it.
I use many java analysis tools to find errors and bugs.
Executing gradle build in eclipse or via command line, during the build process, **taskHistory.bin** changes and grows its dimensions.
This fact triggers the error `java.io.IOException: No space left on device` in Jenkins server.
This happens even with older gradle versions.

### Steps to Reproduce (for bugs)
project example: [projectA.zip](https://github.com/gradle/gradle/files/1950841/projectA.zip)
In the zip, there is a root project with 3 subprojects, if you run from the wrapper `./gradlew build` the file `./.gradle/4.7/taskHistory/taskHistory.bin` is created.
After that, duplicate **ApiClazz.java** in another class, run `./gradlew build` , the file `taskHistory.bin` changes. Do the same with **DriversClazz.java** and the size increases again.
With lots of plugins and many changes, `taskHistory.bin` becomes really big.

### Your Environment
Gradle 4.7
java 8

eclipse oxygen 4.7.3a
buildship eclipse plugin 2.2.1

java analysis tools:
spotless
error prone

**Comment by wolfs** [comment]
This is somewhat related to https://github.com/gradle/gradle/issues/1085, but not very much.

In general, Gradle does not do a very good job in cleanup up its caches. For the task history it is also somewhat hard to decide what to clean up. On running a clean, `taskHistory.bin` becomes pretty useless, since all the tasks will be out of date anyway.

Currently, `taskHistory.bin` will hold the history of the last time any task was executed, but not any more. So if the task names do not change much, the history should also not have stale entries and should grow proportional to the number of files the tasks have as inputs and outputs.

**Comment by regrog** [comment]
> So if the task names do not change much, the history should also not have stale entries and should grow proportional to the number of files the tasks have as inputs and outputs.

Well, my test case shows up that even adding new classes to the plugin increases the size of `taskHistory.bin`

**Comment by wolfs** [comment]
@regrog You are right, we still have some problems cleaning up the task history. Note that it is expected that the task history grows if there are more files as inputs/outputs (i.e. adding a source file), but it should recover the space at some point if files are only renamed/changed.

**Comment by regrog** [comment]
Any improvements about this?

**Comment by lptr** [comment]
FTR, we are investigating moving to H2/MVStore for execution history and other local caches. Doing so will allow us to implement regular cleanups for these caches. (Gradle only has cache cleanup for file-based caches, i.e. where individual entries are represented by entire files.)

**Comment by JamesXNelson** [comment]
what is lost if the executionHistory.bin is deleted?  will it degrade the ability for future builds to skip doing work?

we have some persistent gradle caching, and I'm trying to figure out the right timeline for cleaning out this file. Nightly? weekly? Every run?

**Comment by arielman** [comment]
Any update on this issue?
